scalapact-core/src/main/scala/com/itv/scalapactcore/common/matching/InteractionMatchingPrograms.scala

```diff
@@ -0,0 +1,213 @@
+sealed trait MatchPathOutcome
+case object MatchPathSuccess extends MatchPathOutcome
+case object MatchPathFailed extends MatchPathOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchResponseStageA[A]
+
+case class MatchResponseStatus(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchStatusOutcome]
+case class MatchResponseHeaders(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchHeadersOutcome]
+case class MatchResponseBody(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchBodyOutcome]
+
+
+sealed trait MatchRequestStageA[A]
+
+case class MatchRequestMethod(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchMethodOutcome]
+case class MatchRequestPath(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchPathOutcome]
+case class MatchRequestHeaders(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchHeadersOutcome]
+case class MatchRequestBody(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchBodyOutcome]
+
+case class RequestMatchOutcome(method: MatchMethodOutcome, path: MatchPathOutcome, headers: MatchHeadersOutcome, body: MatchBodyOutcome) {
+  val success: Boolean = {
+    (method, path, headers, body) match {
+      case (MatchMethodSuccess, MatchPathSuccess, MatchHeadersSuccess, MatchBodySuccess) => true
+      case _ => false
+    }
+  }
+
+  val failed: Boolean = !success
+}
+
+case class ResponseMatchOutcome(status: MatchStatusOutcome, headers: MatchHeadersOutcome, body: MatchBodyOutcome) {
+  val success: Boolean = {
+    (status, headers, body) match {
+      case (MatchStatusSuccess, MatchHeadersSuccess, MatchBodySuccess) => true
+      case _ => false
+    }
+  }
+
+  val failed: Boolean = !success
+}
+
+object InteractionMatchingPrograms {
+
+  type MatchResponseStage[A] = Free[MatchResponseStageA, A]
+
+  // Response
+  private def matchResponseStatus(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchStatusOutcome] =
+    Free.liftF(MatchResponseStatus(expected, actual))
+
+  private def matchResponseHeaders(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchHeadersOutcome] =
+    Free.liftF(MatchResponseHeaders(expected, actual))
+
+  private def matchResponseBody(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchBodyOutcome] =
+    Free.liftF(MatchResponseBody(expected, actual))
+
+  def matchResponseProgram(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[ResponseMatchOutcome] =
+    for {
+      s <- matchResponseStatus(expected, actual)
+      h <- matchResponseHeaders(expected, actual)
+      b <- matchResponseBody(expected, actual)
+    } yield ResponseMatchOutcome(s, h, b)
```

Comment:
Here is the program, devoid of actual implementation that describes how to check for a matching response.